plugins/libplugin-pay.c

```diff
@@ -828,9 +828,86 @@ static struct command_result *payment_finished(struct payment *p)
 			 * return a response twice. */
 			p->cmd = NULL;
 			return command_finished(cmd, ret);
-		} else {
-			return command_fail(p->cmd, JSONRPC2_INVALID_REQUEST,
-					    "Not functional yet");
+		} else if (result.failure == NULL || result.failure->failcode < NODE) {
+			/* This is failing because we have no more routes to try */
+			ret = jsonrpc_stream_fail(cmd, PAY_ROUTE_NOT_FOUND,
+						  NULL);
+			json_add_string(
+			    ret, "message",
+			    tal_fmt(cmd,
+				    "Ran out of routes to try after "
+				    "%d attempt%s: see paystatus",
+				    result.attempts,
+				    result.attempts == 1 ? "" : "s"));
+			return command_finished(cmd, ret);
+
+		}  else {
+			struct payment_result *failure = result.failure;
+			assert(failure!= NULL);
+			ret = jsonrpc_stream_fail(cmd, p->result->code,
+						  failure->message);
+
+			json_add_u64(ret, "id", p->result->id);
+
+			json_add_u32(ret, "failcode", result.failure->failcode);
+			json_add_string(ret, "failcodename",
+					result.failure->failcodename);
+			json_add_num(ret, "code", result.failure->code);
+
+			json_object_start(ret, "data");
```

Comment:
That's weird.  jsonrpc_stream_fail() already starts the data obj.  So we return `"error": {"code": 7, "message": "some message", "data": { "id": 1, "failcode": 4103, "failcodename": "temporary_channel_failure", "code": ???, "data": { ...`